CMSgov/mint-app
# Easi 2175/task list locking

# EASI-2175

## Changes and Description

- SubscriptionWrapper
    - Fetches initial state of subscriptions
    - Creates subscription to model plan
    - Receives ws messages and updates subscription state on each message
    - Create SubscriptionContext to be used in application
    - Closes subscription on non-model plan task list sections
- SubscriptionHandler
    - Checks if locations in application are locked, unlocked, can't lock, or already occupying
    - Send mutations to lock and unlock
    - Redirects to locked task list page if page is locked
- Root app Apollo config for handling both http and ws requests
- Locked Task List Page
    - Renders notification page if section is locked
    - Renders error if locking or unlocking failed
- Locked Task List Button
    - Disabled task list button and IconInitial component on locked task list sections
- IT Tools Locked Modal
    - Modal displayed on IT Tools when attempting to redirect to a locked task list section
- Cypress/Unit Testing
    - Created an Apollo client in the Cypress suite to lock sections while running tests to simulate external user
- Backend auth changes to handle websocket authentication

The bread and butter of all the subscription functionality happens in SubscriptionWrapper/index.tsx and SubscriptionHandler/index.tsx

![Screen Shot 2022-09-01 at 9 50 21 AM](https://user-images.githubusercontent.com/95709965/187930810-d0cb3a67-69a3-4943-ac32-b906fefc1575.png)

![Screen Shot 2022-09-01 at 9 51 25 AM](https://user-images.githubusercontent.com/95709965/187930982-91d4cfcb-5df5-4983-8aca-49163fdfe252.png)

![Screen Shot 2022-09-01 at 10 08 15 AM](https://user-images.githubusercontent.com/95709965/187934639-9fd5225d-47ef-4099-8917-238230240973.png)

## Notes

*MINT_ASSESSMENT role currently doesn't have access to lock/unlock (cant enter task list if ONLY has USER_ASSESSMENT), use MINT_USER or both roles for now.  Will possibly address in the future

- There are backend auth changes that were made on this branch as a part of websocket auth
- There are lots of files with spacing adjustments on BE, accounting for lots of touched files
- Refactor of IconInitial component to be used more flexibly
- Moved the collaborator query out of side nav component and into task list to render locked section info

## How to test this change

- Start a model plan as MINT and add Cosmo Kramer as a collaborator
- In another browser (Firefox, etc) log in as Cosmo Kramer (KR14) and go to task list for created model plan
- As KR14 enter General Characteristics task list section
- Verify on MINT's browser than General Characteristics has locked in real time
- On MINT, go to IT Tools page one and click 'Go to question' on the first question
- Verify the modal appears, indicating the locked task list section
- As MINT Attempt to navigate to http://localhost:3005/models/${modelID}/task-list/characteristics
- Verify you are taken to a notification page for locked task list section

Open multiple tabs as the same user and verify you can occupy the same task list section as the same user

Refresh and close tab/browser on one user and verify sections become immediately unlocked on other user browser

Inspect the ws network connection to verify closed connections on any page that isn't part of a model plan or task list. For example the subscription is closed on readonly, home page, cookies, etc.  Connections are only kept open on task list and and task list sections of a model plan


## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
